# Assignment Week 3

Explore, Segment, and Cluster the neighborhoods in the city of Toronto

In [4]:
import requests
import json

import matplotlib.pyplot as plt
import matplotlib.cm as cm
import matplotlib.colors as colors

!pip install beautifulsoup4
from bs4 import BeautifulSoup

import pandas as pd
import numpy as np

print('Libraries imported.')

Libraries imported.


#### 1. Scrape the Wikipedia Page Data.

In [5]:
return_text = requests.get('https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M').text
#print(return_text)
# the table that contains the data we need, start with <table class="wikitable sortable">
soup = BeautifulSoup(return_text)
table = soup.find('table', {'class': 'wikitable sortable'})
print(table)

<table class="wikitable sortable">
<tbody><tr>
<th>Postcode</th>
<th>Borough</th>
<th>Neighbourhood
</th></tr>
<tr>
<td>M1A</td>
<td>Not assigned</td>
<td>Not assigned
</td></tr>
<tr>
<td>M2A</td>
<td>Not assigned</td>
<td>Not assigned
</td></tr>
<tr>
<td>M3A</td>
<td><a href="/wiki/North_York" title="North York">North York</a></td>
<td><a href="/wiki/Parkwoods" title="Parkwoods">Parkwoods</a>
</td></tr>
<tr>
<td>M4A</td>
<td><a href="/wiki/North_York" title="North York">North York</a></td>
<td><a href="/wiki/Victoria_Village" title="Victoria Village">Victoria Village</a>
</td></tr>
<tr>
<td>M5A</td>
<td><a href="/wiki/Downtown_Toronto" title="Downtown Toronto">Downtown Toronto</a></td>
<td><a href="/wiki/Harbourfront_(Toronto)" title="Harbourfront (Toronto)">Harbourfront</a>
</td></tr>
<tr>
<td>M5A</td>
<td><a href="/wiki/Downtown_Toronto" title="Downtown Toronto">Downtown Toronto</a></td>
<td><a href="/wiki/Regent_Park" title="Regent Park">Regent Park</a>
</td></tr>
<tr>
<td>M6A</td>

#### 2. Find and extract table that has the word 'wikitable sortable'

#### 3. Extract data from table and put into a data frame

In [6]:
postcode_rows = []

for table_row in table.findAll('tr'):
    columns = table_row.findAll('td')
    postcode_row = []
    for column in columns:
        postcode_row.append(column.text)
    postcode_rows.append(postcode_row)
    
df=pd.DataFrame(postcode_rows, columns=['PostalCode', 'Borough', 'Neighbourhood'])
df.head(10)

PostalCode           Borough       Neighbourhood
0       None              None                None
1        M1A      Not assigned      Not assigned\n
2        M2A      Not assigned      Not assigned\n
3        M3A        North York         Parkwoods\n
4        M4A        North York  Victoria Village\n
5        M5A  Downtown Toronto      Harbourfront\n
6        M5A  Downtown Toronto       Regent Park\n
7        M6A        North York  Lawrence Heights\n
8        M6A        North York    Lawrence Manor\n
9        M7A      Queen's Park      Not assigned\n

#### 4. Clean up data - replace '\n' and first row that contain 'None'

In [7]:
# remove \n from Neighbourhood
df['Neighbourhood'] = df['Neighbourhood'].replace('\n','', regex=True)

# drop the first row which contain 'None'
df.drop(df.index[0], inplace=True)

df.shape

(288, 3)

In [8]:
df[df['Borough']=='Not assigned'].count()

PostalCode       77
Borough          77
Neighbourhood    77
dtype: int64

In [9]:
#Remove Borough wiht 'Not assigned'
# Assigning df with Borough != 'Not assigned' to df
df = df[df.Borough != 'Not assigned']

# To confirm correct number of rows deleted
df.shape

(211, 3)

In [10]:
#If the neifhbourhood is 'Not assigned', assign its borough name to the neighbouthood
df.loc[df['Neighbourhood'] == 'Not assigned', 'Neighbourhood'] = df['Borough']
df.head(15)

PostalCode           Borough     Neighbourhood
3         M3A        North York         Parkwoods
4         M4A        North York  Victoria Village
5         M5A  Downtown Toronto      Harbourfront
6         M5A  Downtown Toronto       Regent Park
7         M6A        North York  Lawrence Heights
8         M6A        North York    Lawrence Manor
9         M7A      Queen's Park      Queen's Park
11        M9A         Etobicoke  Islington Avenue
12        M1B       Scarborough             Rouge
13        M1B       Scarborough           Malvern
15        M3B        North York   Don Mills North
16        M4B         East York  Woodbine Gardens
17        M4B         East York     Parkview Hill
18        M5B  Downtown Toronto           Ryerson
19        M5B  Downtown Toronto   Garden District

In [11]:
df = df.groupby(["PostalCode", "Borough"], as_index=False).agg(lambda x: ", ".join(x))
df.head(15)

PostalCode      Borough                                      Neighbourhood
0         M1B  Scarborough                                     Rouge, Malvern
1         M1C  Scarborough             Highland Creek, Rouge Hill, Port Union
2         M1E  Scarborough                  Guildwood, Morningside, West Hill
3         M1G  Scarborough                                             Woburn
4         M1H  Scarborough                                          Cedarbrae
5         M1J  Scarborough                                Scarborough Village
6         M1K  Scarborough        East Birchmount Park, Ionview, Kennedy Park
7         M1L  Scarborough                    Clairlea, Golden Mile, Oakridge
8         M1M  Scarborough    Cliffcrest, Cliffside, Scarborough Village West
9         M1N  Scarborough                        Birch Cliff, Cliffside West
10        M1P  Scarborough  Dorset Park, Scarborough Town Centre, Wexford ...
11        M1R  Scarborough                                  Maryvale, Wexford
12        M1S  Scarborough                                          Agincourt
13        M1T  Scarborough            Clarks Corners, Sullivan, Tam O'Shanter
14        M1V  Scarborough  Agincourt North, L'Amoreaux East, Milliken, St...

In [12]:
df.shape

(103, 3)

Finally, there are total 103 rows of records

### End of Part 1

# Part 2 - Get the Latitude and Longitude Coordinates of each Neighborhood.

In [13]:
# Read the cvs file and convert it to a dataframe
import pandas as pd
url='http://cocl.us/Geospatial_data'
df_geo=pd.read_csv(url)
df_geo.columns = ['PostalCode', 'Latitude', 'Longitude']
df_geo.head()

PostalCode   Latitude  Longitude
0        M1B  43.806686 -79.194353
1        M1C  43.784535 -79.160497
2        M1E  43.763573 -79.188711
3        M1G  43.770992 -79.216917
4        M1H  43.773136 -79.239476

In [14]:
df_loc = pd.merge(df, df_geo, on=['PostalCode'], how='inner')
df_loc.head(15)


PostalCode      Borough                                      Neighbourhood  \
0         M1B  Scarborough                                     Rouge, Malvern   
1         M1C  Scarborough             Highland Creek, Rouge Hill, Port Union   
2         M1E  Scarborough                  Guildwood, Morningside, West Hill   
3         M1G  Scarborough                                             Woburn   
4         M1H  Scarborough                                          Cedarbrae   
5         M1J  Scarborough                                Scarborough Village   
6         M1K  Scarborough        East Birchmount Park, Ionview, Kennedy Park   
7         M1L  Scarborough                    Clairlea, Golden Mile, Oakridge   
8         M1M  Scarborough    Cliffcrest, Cliffside, Scarborough Village West   
9         M1N  Scarborough                        Birch Cliff, Cliffside West   
10        M1P  Scarborough  Dorset Park, Scarborough Town Centre, Wexford ...   
11        M1R  Scarborough                                  Maryvale, Wexford   
12        M1S  Scarborough                                          Agincourt   
13        M1T  Scarborough            Clarks Corners, Sullivan, Tam O'Shanter   
14        M1V  Scarborough  Agincourt North, L'Amoreaux East, Milliken, St...   

     Latitude  Longitude  
0   43.806686 -79.194353  
1   43.784535 -79.160497  
2   43.763573 -79.188711  
3   43.770992 -79.216917  
4   43.773136 -79.239476  
5   43.744734 -79.239476  
6   43.727929 -79.262029  
7   43.711112 -79.284577  
8   43.716316 -79.239476  
9   43.692657 -79.264848  
10  43.757410 -79.273304  
11  43.750072 -79.295849  
12  43.794200 -79.262029  
13  43.781638 -79.304302  
14  43.815252 -79.284577

In [15]:
df_loc.shape

(103, 5)

# End of Part 2